src/LanguageServer/Impl/Implementation/WorkspaceManager.cs

```diff
@@ -0,0 +1,74 @@
+﻿using Microsoft.Extensions.FileSystemGlobbing;
+using Microsoft.Extensions.FileSystemGlobbing.Abstractions;
+using Microsoft.PythonTools.Analysis;
+using Microsoft.PythonTools.Analysis.Infrastructure;
+using Microsoft.PythonTools.Parsing;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Python.LanguageServer.Implementation {
+    class WorkspaceManager {
+        private string _rootDir;
+        private string[] _incluideFiles;
+        private string[] _excludeFiles;
+        private IndexingFileParser _indexingFileParser;
+        private Matcher _matcher;
+
+        public WorkspaceManager(string rootDir, string[] includeFiles, string[] excludeFiles, IndexingFileParser indexingFileParser) {
+            _rootDir = rootDir;
+            _incluideFiles = includeFiles;
+            _excludeFiles = excludeFiles;
+            _indexingFileParser = indexingFileParser;
+            if (string.IsNullOrEmpty(_rootDir)) {
+                throw new ArgumentException($"{nameof(rootDir)} null or empty", nameof(rootDir));
+            }
+            InitializeMatcher();
+        }
+
+        private Task<IEnumerable<string>> AsyncDirectoryFilePaths() {
```

Comment:
All async calls must have cancellation support. Ie add `CancellationToken cancellationToken = default` and when task runs, do `cancellationToken.ThrowIfCancellationRequested` in some places that may take time, like long loops or enumerations.